### DISCUSSION TITLE
libpag是否可以在QWidget里面使用

### DISCUSSION THREAD

**Discussion by toby20130333** [Tencent/libpag#240] [Q&A] [answered]
hi
已经看到Qt的例子，做的很棒
想知道是否有QWidgets的例子呢？

**Comment by kevingpqi123** [comment]
暂时没有提供，这里仅仅就是OpenGL环境的提供，可以自己适配下

**Comment by domchen** [comment, marked as answer]
QT官方已经不推荐继续使用QWidget了。我们最早也是基于QWidget实现的，会有一大堆的问题。没有办法做成子线程渲染。后来换成QML就什么问题都没了。建议不要往废弃的坑里跳。：）

**Reply by toby20130333** [reply]
不能吧，QWidgets在桌面客户端开发领域还是会持续很长一段时间的啊，qml只使用做一个较简洁的或者交互简洁的工具吧

**Reply by domchen** [reply]
是我之前描述的不太准确。对标QWidget的框架是Qt Quick，QML只是QtQuick中的一部分。Qt Quick是用来取代老的 QWidget的平行方案，QT官方后续的框架迭代重心都在Qt Quick上了，Qt Quick完全也可以不基于 QML，直接C++编写。它解决了 QWidget 的很多历史设计问题。

**Reply by domchen** [reply]
你如果必须运行在QWidget里，需要自行适配一下。但提醒一下 QWidget 下的 OpenGL 环境几乎不可用，例如 QT 暴露给你的OpenGL 的 context 会污染它本身绘制UI Context。你要是只是用 QWidget 跑一些常规的 UI 程序应该问题还不算大。但是涉及 OpenGL 的环境的时候会出很多的兼容性问题。还是不建议用老的 QWidget 去做 OpenGL 渲染相关的内容。

**Reply by toby20130333** [reply]
@domchen 谢谢提醒，我们目前也只是预研下 这个动画库，看看是否有可行性，但是我们项目需求和功能都是QWidgets，而且也只能是QWidgets
